sim/SPI_v3/components/SPITestHarness.py

```diff
@@ -0,0 +1,208 @@
+'''
+==========================================================================
+SPITestHarness.py
+==========================================================================
+Test harness for sending messages over SPI. For use with SPIMinionAdapterComposite module
+
+Author : Kyle Infantino and Dilan Lakhani
+  Date : Jan 30, 2022
+'''
+
+from pymtl3 import *
+from pymtl3.stdlib.test_utils import config_model_with_cmdline_opts
+
+from math import ceil
+import copy
+
+#=========================================================================
+# TestHarness
+#=========================================================================
+
+class SPITestHarness( object ):
+
+  #-----------------------------------------------------------------------
+  # constructor
+  #-----------------------------------------------------------------------
+ 
+  def __init__( s, DESIGN, num_components, spi_bits, cmdline_opts ):
+    """
+    DESIGN: instantiation of RTL design
+    num_components: number of components that can be addressed
+    spi_bits: number of bits in an spi packet
+    """
```

Comment:
We typically don't use docstrings. We normally put the description before the class definition.

Reply:
Moved to before init declaration